rj_gameplay/rj_gameplay/basic_play_selector.py

```diff
@@ -1,89 +1,271 @@
+        self.field_loc = self.__calc_field_loc(world_state, game_info)
+        self.ball_pos = self.__calc_ball_pos(world_state, game_info)
+        self.is_pileup = self.__calc_pileup(world_state, game_info)
+
+    @staticmethod
+    def __calc_field_loc(
+        world_state: stp.rc.WorldState, game_info: stp.rc.GameInfo
+    ) -> FieldLoc:
+        """Computes the current location of the ball.
+        :param world_state:
+        :param game_info:
+        :return: The current FieldLoc.
+        """
+        ball_pos: np.ndarray = world_state.ball.pos
+
+        field_len: float = world_state.field.length_m
+        midfield: float = field_len / 2
+
+        if ball_pos[1] <= midfield:
+            return FieldLoc.DEFEND_SIDE
+        elif ball_pos[1] > midfield:
+            return FieldLoc.ATTACK_SIDE
+        else:
+            return FieldLoc.MIDFIELD
+
+    @staticmethod
+    def __calc_ball_pos(
+        world_state: stp.rc.WorldState, game_info: stp.rc.GameInfo
+    ) -> BallPos:
+        """Computes the current ball possession.
+        :param world_state:
+        :param game_info:
+        :return: The current BallPos.
+        """
+        for our_bot in world_state.our_robots:
+            if (
+                our_bot.has_ball_sense
+                or np.linalg.norm(
+                    np.array(world_state.ball.pos) - np.array(our_bot.pose[0:2])
+                )
+                < POSSESS_MIN_DIST
+            ):
+                return BallPos.OUR_BALL
+        for their_bot in world_state.their_robots:
+            if (
+                np.linalg.norm(
+                    np.array(world_state.ball.pos) - np.array(their_bot.pose[0:2])
+                )
+                < POSSESS_MIN_DIST
+            ):
+                return BallPos.THEIR_BALL
+        if np.linalg.norm(world_state.ball.vel) > MIN_PASS_SPEED:
+            for our_bot in world_state.our_robots:
+                ball_to_bot = np.array(world_state.ball.pos) - np.array(
+                    our_bot.pose[0:2]
+                )
+                ball_to_bot_unit = ball_to_bot / np.linalg.norm(ball_to_bot)
+                ball_dir = world_state.ball.vel / np.linalg.norm(world_state.ball.vel)
+                dot = abs(np.dot(ball_to_bot_unit, ball_dir))
+                if dot > 0.7:
+                    return BallPos.OUR_BALL
+            for their_bot in world_state.their_robots:
+                ball_to_bot = np.array(world_state.ball.pos) - np.array(
+                    their_bot.pose[0:2]
+                )
+                ball_to_bot_unit = ball_to_bot / np.linalg.norm(ball_to_bot)
+                ball_dir = world_state.ball.vel / np.linalg.norm(world_state.ball.vel)
+                dot = abs(np.dot(ball_to_bot_unit, ball_dir))
+                if dot > 0.7:
+                    return BallPos.THEIR_BALL
+        # TODO(1578): Re-implement actual previous logic here.
+
+        return BallPos.FREE_BALL
+
+    @staticmethod
+    def __calc_pileup(
```

Comment:
I think it might be nicer if this was included in the calc_ball_pos with a fourth pileup state.